pkg/controller/nodeport-proxy/envoymanager/resources.go

```diff
@@ -27,21 +27,259 @@ import (
 	"github.com/golang/protobuf/ptypes"
 	"github.com/golang/protobuf/ptypes/wrappers"
 	"github.com/pkg/errors"
+	"google.golang.org/protobuf/types/known/anypb"
 	corev1 "k8s.io/api/core/v1"
 
+	//	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+
+	envoyaccesslogv3 "github.com/envoyproxy/go-control-plane/envoy/config/accesslog/v3"
 	envoyclusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
 	envoycorev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
 	envoyendpointv3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
 	envoylistenerv3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
 	envoyroutev3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	envoylistenerlogv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/access_loggers/file/v3"
 	envoyhealthv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/health_check/v3"
 	envoyhttpconnectionmanagerv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
 	envoytcpfilterv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/tcp_proxy/v3"
 	envoycachetype "github.com/envoyproxy/go-control-plane/pkg/cache/types"
 	envoycachev3 "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
 	envoywellknown "github.com/envoyproxy/go-control-plane/pkg/wellknown"
 )
 
+const (
+	HTTPSAltPort  = 8443
+	HTTPProxyPort = 8080
+	UpgradeType   = "CONNECT"
+)
+
+var stdoutAccessLog *anypb.Any
+
+func init() {
+	var err error
+	f := &envoylistenerlogv3.FileAccessLog{
+		Path: "/dev/stdout",
+	}
+
+	stdoutAccessLog, err = ptypes.MarshalAny(f)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func makeAccessLog() []*envoyaccesslogv3.AccessLog {
+	accessLog := []*envoyaccesslogv3.AccessLog{
+		{
+			Name: wellknown.FileAccessLog,
+			ConfigType: &envoyaccesslogv3.AccessLog_TypedConfig{
+				TypedConfig: stdoutAccessLog,
+			},
+		},
+	}
+	return accessLog
+}
+
+func makeSNIFilterChainMatch(service *corev1.Service) *envoylistenerv3.FilterChainMatch {
+	filterChainMatch := &envoylistenerv3.FilterChainMatch{
+		// New annotation used for the new expose strategy. The "expose" one will stay
+		// for the default expose.
+		ServerNames:       []string{service.ObjectMeta.Annotations["nodeport-proxy.k8s.io/domainname"]},
```

Comment:
in theory, the Annotations map should not be `nil` but I would add a check, also the specific annotation may be not present, in this case we are getting an empty string.

Maybe you should just add `serverName` as a function parameter. It will be the responsibility of the caller not to call this function if the annotation is not present.